sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriteOutput.scala

```diff
@@ -0,0 +1,478 @@
+      allColumns = plan.output,
+      partitionColumns = partitionColumns,
+      nonPartitionColumns = dataColumns,
+      isAppend = isAppend,
+      path = outputPath,
+      outputFormatClass = job.getOutputFormatClass,
+      bucketSpec = bucketSpec)
+
+    val serializableHadoopConf = new SerializableConfiguration(hadoopConf)
+
+    SQLExecution.withNewExecutionId(sparkSession, queryExecution) {
+
+      // This call shouldn't be put into the `try` block below because it only initializes and
+      // prepares the job, any exception thrown from here shouldn't cause abortJob() to be called.
+      val committer = setupDriverCommitter(hadoopConf, job, outputPath, isAppend)
+
+      try {
+        sparkSession.sparkContext.runJob(queryExecution.toRdd,
+          (taskContext: TaskContext, iter: Iterator[InternalRow]) => {
+            executeTask(context, serializableHadoopConf.value, taskContext, iter)
+          })
+
+        committer.commitJob(job)
+        logInfo(s"Job ${job.getJobID} committed.")
+
+        refreshFunction()
+      } catch { case cause: Throwable =>
+        logError(s"Aborting job ${job.getJobID}.", cause)
+        committer.abortJob(job, JobStatus.State.FAILED)
+
+        throw new SparkException("Job aborted.", cause)
+      }
+    }
+  }
+
+  class WriteJobDescription(
+      val outputWriterFactory: OutputWriterFactory,
+      val allColumns: Seq[Attribute],
+      val partitionColumns: Seq[Attribute],
+      val nonPartitionColumns: Seq[Attribute],
+      val isAppend: Boolean,
+      val path: Path,
+      val outputFormatClass: Class[_ <: OutputFormat[_, _]],
+      val bucketSpec: Option[BucketSpec])
+    extends Serializable
+
+  def executeTask(
+      description: WriteJobDescription,
+      hadoopConf: Configuration,
+      taskContext: TaskContext,
+      iterator: Iterator[InternalRow]): Unit = {
+    // Setup IDs
+    val jobId = SparkHadoopWriter.createJobID(new Date, taskContext.stageId())
+    val taskId = new TaskID(jobId, TaskType.MAP, taskContext.partitionId())
+    val taskAttemptId = new TaskAttemptID(taskId, taskContext.attemptNumber())
+
+    // Set up the configuration object
+    hadoopConf.set("mapred.job.id", jobId.toString)
+    hadoopConf.set("mapred.tip.id", taskAttemptId.getTaskID.toString)
+    hadoopConf.set("mapred.task.id", taskAttemptId.toString)
+    hadoopConf.setBoolean("mapred.task.is.map", true)
+    hadoopConf.setInt("mapred.task.partition", 0)
+
+    val taskAttemptContext = new TaskAttemptContextImpl(hadoopConf, taskAttemptId)
+    val committer = newOutputCommitter(
+      description.outputFormatClass, taskAttemptContext, description.path, description.isAppend)
+    committer.setupTask(taskAttemptContext)
+
+    if (description.partitionColumns.isEmpty && description.bucketSpec.isEmpty) {
+      executeSingleDirectoryWriteTask(
+        description,
+        taskAttemptContext,
+        committer,
+        iterator)
+    } else {
+      executeDynamicPartitionWriteTask(
```

Comment:
note to self: it'd be better if we can move the task commit and abort logic into this function